src/system/SystemPacketBuffer.cpp

```diff
@@ -681,7 +681,7 @@ PacketBufferHandle PacketBufferHandle::CloneData() const
 
     for (PacketBuffer * original = mBuffer; original != nullptr; original = original->ChainedBuffer())
     {
-        uint16_t originalDataSize     = original->MaxDataLength();
+        uint16_t originalDataSize     = original->DataLength();
```

Comment:
But if we want the clone's sizing the match the original buffer, don't we need to in fact use `MaxDataLength` for the data size?

That said, my understanding is that the source buffer absolutely must be big enough to hold `MaxDataLength` + `ReservedSize`.  Under what conditions would that not hold?

Reply:
yes, it would be nice to match the allocated size but the problem is, the buffer does not preserve original allocated size, at least not in lwip's pbuf case.

Reply:
What is the MaxDataLength for this case?

We have code in the tree that generally assumes it's OK to write into a packetbuffer up to the MaxDataLength, so if the MaxDataLength is wrong we should fix that....

Reply:
so, here's the full ASAN message i revert this change:

https://gist.githubusercontent.com/rojer/19eedce618e6b1a9b3d4974962fb3a3a/raw/de35e7159f3cb69157fedf69190c9fe7743e283d/gistfile1.txt

a 152-byte-packet is being cloned and copied into a 1520 one and memcpy goes off into the weeds copying the extra garbage. given what you said, maybe the right approach is to allocate `MaxDataLength()` but only copy `DataLength()` bytes. though it looks like a waste of space. with my fix i don't get in trouble with ASAN anymore, so it seems fine. i mean, the underlying source pbuf is already smaller than MaxDataLength, so we are kind of already in breach of that assumption...
anyway, i'm deferring to you here on what the appropriate course of action should be.

Reply:
> a 152-byte-packet is being cloned and copied

But what is the MaxDataLength() of that 152-byte thing?  Does that buffer have `PBUF_STRUCT_DATA_CONTIGUOUS` testing true or false?  I assume this is using a version of Matter where the value of `MaxDataLength` does in fact depend on `PBUF_STRUCT_DATA_CONTIGUOUS`?

> with my fix i don't get in trouble with ASAN anymore, so it seems fine

It seems like it's sweeping under the rug the issue that MaxDataLength claims a completely wrong value, which will cause problems if someone ever tries to write to that buffer....

> the underlying source pbuf is already smaller than MaxDataLength

That should never happen.  If that happens, we need to fix it to not happen.  The question is: why is it happening in this case?

Reply:
ok, on closer inspection it seems that i'm trying to use a mode that is not currently supported: in order to reduce static memory usage i'm trying to use as-needed, dynamically allocated pbufs. to that end, i'm setting

```
#define CHIP_SYSTEM_CONFIG_PACKETBUFFER_LWIP_PBUF_TYPE PBUF_RAM
#define CHIP_SYSTEM_CONFIG_PACKETBUFFER_POOL_SIZE 0
```

what i really want is `CHIP_SYSTEM_PACKETBUFFER_FROM_CHIP_HEAP` _and_ pbufs.
let me ponder this a bit more.

Reply:
ok, so it all makes sense now. this was my misunderstanding of how CHIP uses pbufs - my apologies as i get myself more acquainted with CHIP.
looks like CHIP relies on pool-allocated pbufs, and cannot handle `CHIP_SYSTEM_CONFIG_PACKETBUFFER_LWIP_PBUF_TYPE` other than `PBUF_POOL` (specifically `PBUF_RAM`) exactly because there's no way to track allocated size and correctly return AllocSize. non-lwip implementation of pbuf does add alloc_size member for this case.
however, using `PBUF_RAM` for outgoing packets is very common and in fact lwip's guide to pbuf types discourages use of PBUF_POOL for outbound packets ([here](https://github.com/lwip-tcpip/lwip/blob/84fde1ebbfe35b3125fc2d89b8a456cbacf148e9/src/include/lwip/pbuf.h#L165-L167)):
```
      Don't use this for TX, if the pool becomes empty e.g. because of TCP queuing,
      you are unable to receive TCP acks!
```

i tried to modify `SystemPacketBuffer` to support `PBUF_RAM` type pbufs but it proved too much for a quick stab, it's a `ENOTIME` for me at the moment. however, i think it's a good idea to eventually do it. two ideas for tracking alloc size and making it happen:
1. add a field via [LWIP_PBUF_CUSTOM_DATA](https://github.com/lwip-tcpip/lwip/blob/84fde1ebbfe35b3125fc2d89b8a456cbacf148e9/src/include/lwip/pbuf.h#L224). we cannot define `LWIP_PBUF_CUSTOM_DATA` ourselves, that belongs to user's lwipopts, but we can require that user adds `alloc_size` for us and ensure `LWIP_PBUF_CUSTOM_DATA` is defined.
2. since we control the allocation size we can add extra `LWIP_MEM_ALIGN_SIZE(sizeof uint16_t)` bytes on top and store our `alloc_size` there. may need some fiddling with the payload pointer too.

@bzbarsky-apple let me know if you think it's a good idea, i'll file a separate issue.

Reply:
Using `LWIP_PBUF_CUSTOM_DATA` to store  `alloc_size` to enable `PBUF_RAM` is better than my previous PR.
Hope the suggestion to enable `PBUF_RAM`  from @rojer can be accepted.

Reply:
@tcarmelveilleux thoughts on https://github.com/project-chip/connectedhomeip/pull/29171#discussion_r1322226606. Does the custom data approach seem better than the things we considered in https://github.com/project-chip/connectedhomeip/pull/28495?

Reply:
Well it assumes that nobody has configured lwip to already use custom data in a way that is not visible to us. In that case it would impact the users if we start making use of it. Overall I think we are stuck in a loop of brittle with what is here. This is why I was wondering if it wouldn't be a better idea to find a way to never rely directly on the pbuf handles and improve PacketBuffer to not descend from pbuf directly. When a packet comes in, we don't care as much about how much space I'd allocated since start big and take parts away. When we create a new packet buffer, that's when all the laying out is important.

Reply:
@tcarmelveilleux the user can always add the necessary field to whatever value `LWIP_PBUF_CUSTOM_DATA` already has. can call it `chip_alloc_size` or whatever, to make it distinct.

either way, i see that maintainers are well aware of this problem. i think i'll keep this hack local, it seems to do the job without breaking things, but it's clearly not the solution. thanks everyone for the discussion, closing the PR.